### PR TITLE
Concept support (minimal functionality)

### PR DESCRIPTION
**Pull request by waitingtocompile**
This brings concepts up to the  bare minimum functionality, with a few notable limitations

1 - concept-constraints on templates are not recorded in any way, the data is lost currently
2 - concepts themselves behave oddly, due to limitations in libclang we can't actually enumerate their template parameters like other templates
3 - codegen can't generate code with concept constraints, and can't generate concepts in general.

Issues 1 and 3 I hope to resolve in the future as I continue working on this. Issue 2 is going to be dependant on libclang properly exposing concept definitions.

### CODE REVIEW

**Comment by waitingtocompile** [comment]
Concept constraint information is now being partially recorded. In-parameter constraints are recorded and generated correctly, however constraints between the template and the actual entity are not, as I've found them difficult to parse reliably

Concepts themselves are now fully generatable. In general I'm really running up against the limits of what information I can actually reliably extract given the lack of libclang support for concepts. Still, it's almost all of the way to the functionality that I want for my purposes, so unless I can figure out a solution for parsing constraints, I'm going to call this "done" beyond any stylistic/formatting changes you want.

Tests are passing properly on my machine, sadly the CI server seems to be running a fairly old version of clang.

File: include/cppast/cpp_concept.hpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    cpp_token_string parameters_;
+
+    std::unique_ptr<cpp_expression> expression_;
+};
+
+
+} // namespace cppast
+
+#endif
```
Comment (foonathan) [inline include/cppast/cpp_concept.hpp:74]: Please add a trailing newline here.

File: include/cppast/cpp_concept.hpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+private:
+    cpp_concept(std::string name) 
+        : cpp_entity(std::move(name)), parameters_({})
+    {}
+
+    cpp_entity_kind do_get_entity_kind() const noexcept override;
+
+    cpp_token_string parameters_;
+
```
Comment (foonathan) [inline include/cppast/cpp_concept.hpp:67]: Nitpick, remove the empty line between the members.

File: src/cpp_concept.cpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+cpp_entity_kind cpp_concept::do_get_entity_kind() const noexcept
+{
+    return cpp_entity_kind::concept_t;
+}
+
+std::unique_ptr<cpp_concept> cpp_concept::builder::finish(const cpp_entity_index& idx, cpp_entity_id id)
+{
+    idx.register_definition(id, type_safe::ref(*concept_));
+    return std::move(concept_);
+}
```
Comment (foonathan) [inline src/cpp_concept.cpp:24]: Trailing newline here as well.

File: src/libclang/concept_parser.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    builder.set_expression(
+        parse_raw_expression(context, stream, stream.end() - 1,
+                             cpp_builtin_type::build(cpp_builtin_type_kind::cpp_bool)));
+
+    builder.set_parameters(std::move(params));
+
+    return builder.finish(*context.idx, detail::get_entity_id(cur));
+}
```
Comment (foonathan) [inline src/libclang/concept_parser.cpp:67]: Trailing newline here as well.

File: src/libclang/concept_parser.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+std::unique_ptr<cpp_entity> detail::try_parse_cpp_concept(const detail::parse_context& context,
+                                                          const CXCursor&              cur)
+{
+    DEBUG_ASSERT(cur.kind == CXCursor_UnexposedDecl, detail::assert_handler{});
+
+    detail::cxtokenizer    tokenizer(context.tu, context.file, cur);
+    detail::cxtoken_stream stream(tokenizer, cur);
+
+    if (!detail::skip_if(stream, "template"))
+    {
```
Comment (foonathan) [inline src/libclang/concept_parser.cpp:22]: Nitpick: please remove braces around if statements that just contain return.

File: src/libclang/template_parser.cpp [hunk 6]
```diff
@@ -42,22 +42,66 @@ type_safe::optional<typename TemplateT::builder> get_builder(const detail::parse
...
+    detail::cxtoken_stream stream(tokenizer, parent);
+    
+    detail::cxtoken_iterator found_start = detail::find_sequence(stream, target_range_start, target_range_end);
+    if(found_start == stream.end())
+    {
+        return detail::to_string(stream, stream.cur() + 1);
+    }
+    stream.set_cur(found_start);
+    stream.bump();
+    if (stream.peek() == "<")
```
Comment (foonathan) [inline src/libclang/template_parser.cpp:61]: Braces here as well.

File: src/libclang/cxtokenizer.cpp [hunk 7]
```diff
@@ -532,6 +532,39 @@ void detail::skip_brackets(detail::cxtoken_stream& stream)
...
+                failed = true;
+            }
+            else
+            {
+                ++search_iter;
+                ++seq_iter;
+            }
+        }
+        if(!failed)
+        {
```
Comment (foonathan) [inline src/libclang/cxtokenizer.cpp:559]: Braces here as well.

File: src/libclang/cxtokenizer.cpp [hunk 8]
```diff
@@ -532,6 +532,39 @@ void detail::skip_brackets(detail::cxtoken_stream& stream)
...
+    {
+        detail::cxtoken_iterator search_iter = search_start;
+        detail::cxtoken_iterator seq_iter    = start;
+        bool                     failed      = false;
+
+        while(!failed && search_iter != stream.end() && seq_iter != end)
+        {
+            if(search_iter->value() != seq_iter->value()
+                || search_iter->kind() != seq_iter->kind())
+            {
```
Comment (foonathan) [inline src/libclang/cxtokenizer.cpp:549]: (Keep the braces here for symmetry)

File: test/cpp_concept.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        else if (tfunc.name() == "f2" || tfunc.name() == "f3")
+        {
+            REQUIRE(static_cast<const cpp_template_type_parameter&>(*tfunc.parameters().begin())
+                        .keyword()
+                    == cpp_template_keyword::concept_contraint);
+        }
+    });
+
+    REQUIRE(count == 3u);
+}
```
Comment (foonathan) [inline test/cpp_concept.cpp:82]: And trailing newline here as well.

**Review by foonathan — APPROVED** [review]
Thank you very much! I just have minor style nitpicks.

**Comment by foonathan** [comment]
> Tests are passing properly on my machine, sadly the CI server seems to be running a fairly old version of clang.

You can use a check for the clang version in the test, e.g. as done here: https://github.com/foonathan/cppast/blob/f3e399573f0dc192625de083afdab05f56d8cba3/test/cpp_function.cpp#L278

I'll see whether I can update CI.

**Comment by foonathan** [comment]
Again, thank you very much.

**Pull request merged**